### PR TITLE
ProcessInfo: save atomically

### PR DESCRIPTION
**Pull request by skshetry**
Trying to fix https://github.com/iterative/dvc/issues/8570#issue-1450769780.

### CODE REVIEW

File: src/dvc_task/proc/process.py [hunk 1]
```diff
@@ -50,12 +51,17 @@ def asdict(self) -> Dict[str, Any]:
...
+        with tempfile.NamedTemporaryFile(
+            mode="w",
+            encoding="utf-8",
+            dir=directory,
+            prefix=f"{file}.",
+            suffix=".tmp",
+            delete=False,
+        ) as tmp:
+            json.dump(self.asdict(), tmp)
+        os.replace(tmp.name, filename)
```
Comment (skshetry) [inline src/dvc_task/proc/process.py:64]: How likely is it that this file gets overwritten? Is it only written once?

In Windows, `os.replace` may fail with PermissionError, if there are open file handles to destination, so for making it robust, we may need to retry.

File: src/dvc_task/proc/process.py [hunk 2]
```diff
@@ -50,12 +51,17 @@ def asdict(self) -> Dict[str, Any]:
 
     def dump(self, filename: str) -> None:
```
Comment (skshetry) [inline src/dvc_task/proc/process.py:52]: In `load()` above, do we need to serialize loading with `lock`/`unlock`? Are there any side-effects from `load()`? Is it possible to get rid of it?

**Comment by skshetry** [comment]
ping @karajan1001 @pmrowla.

File: src/dvc_task/proc/process.py (hunk 1 above)
Comment (karajan1001) [reply src/dvc_task/proc/process.py:64]: Very often, in `manager.py` we will update the process info via `__setitem__` if there are some status updates on queue tasks.

Comment (skshetry) [reply src/dvc_task/proc/process.py:64]: @karajan1001, can you show me where we overwrite this? And when? And what do you mean by "very often", like how frequently that happens?

Apologies for asking too many questions, I'm still trying to understand the codebase. :)

Comment (skshetry) [reply src/dvc_task/proc/process.py:64]: Although `os.replace()` can be fixed separately.

Comment (karajan1001) [reply src/dvc_task/proc/process.py:64]: OK, I'm sorry looking into it I found that I had mixed `ProcessInfo` with `ExecutorInfo`, here the `ProcessInfo` only dumps in https://github.com/iterative/dvc-task/blob/59caee678957668ad52cada7b78fc1f3ef3fa4b7/src/dvc_task/proc/process.py#L203

**Review by karajan1001 — APPROVED** [review]

**Pull request merged**